### PR TITLE
br: a more straight forward operator interface

### LINKED ISSUE #5699: pitr: add support to log backup subcommand of status/pause/resume
## Feature Request

**Is your feature request related to a problem? Please describe:**

**Describe the feature you'd like:**

currently, tidb operator does not support status/pause/resume log backup subcommands, while the pitr kernel supports them. br operator needs to expose the interfaces.

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by RidRisR**
### What problem does this PR solve?

Closes #5699 

1. Use a more straight forward operator interface to replace the old one.
2. support pause and resume pitr task

### What is changed and how does it work?

Now, the crd to control log backup would loops like
```spec:
  backupMode: log
  logSubcommand: log-start/log-stop/log-pause
  logTruncateUntil: "xxxxxxxxx"
```
(If no subcommand is provided, the default value is `log-start`)

If the task is not created before:
It could transfer to `running` state using `log-start`

If the task is running:
It could transfer to `paused` state using `log-pause`
It could transfer to `stop` state using `log-stop`

If the task is paused:
It could transfer to `running` state using `log-start`
It could transfer to `stop` state using `log-stop`

If the task is stopped:
This is the end state and can't transfer state again

As long as the task is created, user could set `logTruncateUntil` and truncate would be sync after subcommand is done

截屏2024-09-02 17 50 46

### Code changes

- [ ] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [ ] Unit test 
- [ ] E2E test 
- [x] Manual test 
- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [x] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
1. Use a more straight forward operator interface to replace the old one.
Now, the crd to control log backup would loops like
`spec:
  backupMode: log
  logSubcommand: `log-start`/`log-stop`/`log-pause`
  logTruncateUntil: "xxxxxxxxx"
`
(If no subcommand is provided, the default value is `log-start`)

2. Support pause and resume pitr task
Now, users could pause and resume pitr task by using `log-pause` and `log-start`
```

### CODE REVIEW

File: pkg/apis/pingcap/v1alpha1/backup.go [hunk 1]
```diff
@@ -321,13 +337,21 @@ func ParseLogBackupSubcommand(backup *Backup) LogSubCommandType {
...
+	switch backup.Spec.LogSubcommand {
+	case "log-start":
+		if IsLogBackupAlreadyPaused(backup) {
+			return LogResumeCommand
+		}
+		return LogStartCommand
+	case "log-stop":
 		return LogStopCommand
-	}
-	if backup.Spec.LogTruncateUntil != "" {
```
Comment (csuzhangxc) [inline pkg/apis/pingcap/v1alpha1/backup.go:327]: Is it needed to compatible with the old fields?

Comment (RidRisR) [reply pkg/apis/pingcap/v1alpha1/backup.go:327]: No, user shouldn't mix up these two interface to avoid mistakes

Comment (csuzhangxc) [reply pkg/apis/pingcap/v1alpha1/backup.go:327]: So the old existing Backup user MUST update their pipeline to use these new fields, right?

Comment (RidRisR) [reply pkg/apis/pingcap/v1alpha1/backup.go:327]: Yes, we don't want handle corner case like:
````
Spec:
    logSubcommand: log-start
    logStop: true
```
BTW, I am fresh to operator. Would there be any problem?
I tested updating operator online while a pitr task is running.
Seems there wouldn't be any problem.

Comment (WangLe1321) [reply pkg/apis/pingcap/v1alpha1/backup.go:327]: We need to keep compatible with the old fields, because tidbcloud depends on current implementation. I think we can make old fields have higher priority. For example:
```
if backup.Spec.LogStop
    return LogStopCommand
if backup.Spec.LogTruncateUntil != "" && backup.Spec.LogTruncateUntil != backup.Status.LogSuccessTruncateUntil
    return LogTruncateCommand

switch backup.Spec.LogSubcommand
...

// default command is log start
return LogStartCommand
```

File: pkg/apis/pingcap/v1alpha1/backup.go [hunk 2]
```diff
@@ -201,6 +201,22 @@ func IsBackupScheduled(backup *Backup) bool {
...
 
+// HaveTruncateUntil returns true if a Backup has truncate until set
+func HaveTruncateUntil(backup *Backup) bool {
+	if backup.Spec.Mode != BackupModeLog {
+		return false
+	}
+	return backup.Spec.LogTruncateUntil != ""
+}
+
+func IsBackupPaused(backup *Backup) bool {
```
Comment (WangLe1321) [inline pkg/apis/pingcap/v1alpha1/backup.go:212]:
```suggestion
func IsLogBackupPaused(backup *Backup) bool {
```

File: pkg/apis/pingcap/v1alpha1/backup.go [hunk 3]
```diff
@@ -201,6 +201,22 @@ func IsBackupScheduled(backup *Backup) bool {
...
+	if backup.Spec.Mode == BackupModeLog {
+		return IsLogBackupSubCommandOntheCondition(backup, BackupPaused)
```
Comment (WangLe1321) [inline pkg/apis/pingcap/v1alpha1/backup.go:214]: Seems log backup sub command condition has no `BackupPaused` status

File: pkg/controller/backup/backup_controller.go [hunk 4]
```diff
@@ -230,6 +230,10 @@ func (c *Controller) updateBackup(cur interface{}) {
 		return
 	}
 
+	if newBackup.Spec.Mode == v1alpha1.BackupModeLog && newBackup.Spec.LogSubcommand == string(v1alpha1.LogStartCommand) && v1alpha1.IsBackupPaused(newBackup) {
+		newBackup.Spec.LogSubcommand = string(v1alpha1.LogResumeCommand)
```
Comment (WangLe1321) [inline pkg/controller/backup/backup_controller.go:234]: I don't know why add this hack logic. We shouldn't modify the `spec` fields even the user inputs wrong value.

File: pkg/apis/pingcap/v1alpha1/backup.go [hunk 5]
```diff
@@ -201,6 +201,22 @@ func IsBackupScheduled(backup *Backup) bool {
...
+	}
+	_, condition := GetBackupCondition(&backup.Status, BackupPaused)
+	return condition != nil && condition.Status == corev1.ConditionTrue
```
Comment (WangLe1321) [inline pkg/apis/pingcap/v1alpha1/backup.go:217]: This only means the log backup is ever paused, not means it is paused. I guess you want to check if the log backup is paused. Only `backup.Status.Phase` represents the current status.

File: pkg/apis/pingcap/v1alpha1/backup.go [hunk 6]
```diff
@@ -392,3 +416,13 @@ func IsLogBackupAlreadyTruncate(backup *Backup) bool {
 func IsLogBackupAlreadyStop(backup *Backup) bool {
 	return backup.Spec.Mode == BackupModeLog && backup.Status.Phase == BackupStopped
 }
+
+// IsLogBackupAlreadyStop return whether log backup has already paused.
```
Comment (WangLe1321) [inline pkg/apis/pingcap/v1alpha1/backup.go:420]:
```suggestion
// IsLogBackupAlreadyPaused return whether log backup has already paused.
```

File: pkg/apis/pingcap/v1alpha1/backup.go [hunk 7]
```diff
@@ -392,3 +416,13 @@ func IsLogBackupAlreadyTruncate(backup *Backup) bool {
...
+func IsLogBackupAlreadyPaused(backup *Backup) bool {
+	return backup.Spec.Mode == BackupModeLog && backup.Status.Phase == BackupPaused
+}
+
+// IsLogBackupAlreadyRestart return whether log backup has already resumed.
```
Comment (WangLe1321) [inline pkg/apis/pingcap/v1alpha1/backup.go:425]:
```suggestion
// IsLogBackupAlreadyRunning return whether log backup has already resumed.
```

**Comment by BornChanger** [comment]
For the state machine figure, please also add the scenarios where the state change due to gc-ttl

File: pkg/apis/pingcap/v1alpha1/backup.go (hunk 3 above)
Comment (RidRisR) [reply pkg/apis/pingcap/v1alpha1/backup.go:214]: But it do have, after adding the pause stage and command in this PR

File: pkg/controller/backup/backup_controller.go (hunk 4 above)
Comment (RidRisR) [reply pkg/controller/backup/backup_controller.go:234]: deleted

File: pkg/apis/pingcap/v1alpha1/backup.go (hunk 5 above)
Comment (RidRisR) [reply pkg/apis/pingcap/v1alpha1/backup.go:217]: deleted

File: pkg/apis/pingcap/v1alpha1/backup.go (hunk 2 above)
Comment (RidRisR) [reply pkg/apis/pingcap/v1alpha1/backup.go]: deleted

File: pkg/apis/pingcap/v1alpha1/backup.go (hunk 1 above)
Comment (RidRisR) [reply pkg/apis/pingcap/v1alpha1/backup.go:327]: Seems `backup.Spec.LogTruncateUntil != backup.Status.LogSuccessTruncateUntil` is unnecessary
The user should be allowed to truncate multiple times.

**Comment by RidRisR** [comment]
> For the state machine figure, please also add the scenarios where the state change due to gc-ttl. @RidRisR

we can't modify advancer time out setting now. therefore, the gc-ttl will only time out after 24 hours, it seems too long for a test.

File: cmd/backup-manager/app/backup/manager.go [hunk 8]
```diff
@@ -582,6 +616,37 @@ func (bm *Manager) stopLogBackup(ctx context.Context, backup *v1alpha1.Backup) (
...
 }
 
+// pauseLogBackup stops log backup.
+func (bm *Manager) pauseLogBackup(ctx context.Context, backup *v1alpha1.Backup) (*controller.BackupUpdateStatus, string, error) {
+	started := time.Now()
+
+	// change Prepare to Running before real backup process start
+	if err := bm.StatusUpdater.Update(backup, &v1alpha1.BackupCondition{
+		Command: v1alpha1.LogPauseCommand,
+		Type:    v1alpha1.BackupRunning,
```
Comment (csuzhangxc) [inline cmd/backup-manager/app/backup/manager.go:626]: Should we need to set `Paused` here?

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 9]
```diff
@@ -2177,6 +2179,9 @@ type BackupSpec struct {
 	// Default is current timestamp.
 	// +optional
 	CommitTs string `json:"commitTs,omitempty"`
+	// Subcommand is the subcommand for BR, such as start, stop, pause etc.
+	// +optional
+	LogSubcommand string `json:"logSubcommand,omitempty"`
```
Comment (csuzhangxc) [inline pkg/apis/pingcap/v1alpha1/types.go:2184]: How about mark it with `+kubebuilder:validation:Enum`?

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 10]
```diff
@@ -2397,6 +2404,12 @@ const (
 	LogTruncateCommand LogSubCommandType = "log-truncate"
 	// LogStopCommand is the stop command of log backup.
 	LogStopCommand LogSubCommandType = "log-stop"
+	// LogPauseCommand is the pause command of log backup.
+	LogPauseCommand LogSubCommandType = "log-pause"
+	// LogResumeCommand is the resume command of log backup.
+	LogResumeCommand LogSubCommandType = "log-resume"
+	// LogUnknownCommand is the unknown command of log backup.
+	LogUnknownCommand LogSubCommandType = "log-unknown"
```
Comment (csuzhangxc) [inline pkg/apis/pingcap/v1alpha1/types.go:2412]: is `log-unknown` set by users? or just an invalid value.

File: cmd/backup-manager/app/backup/manager.go (hunk 8 above)
Comment (RidRisR) [reply cmd/backup-manager/app/backup/manager.go:626]: The running state here means the `command` is running rather than the `log backup` is running.

File: pkg/apis/pingcap/v1alpha1/types.go (hunk 10 above)
Comment (RidRisR) [reply pkg/apis/pingcap/v1alpha1/types.go:2412]: The latter one, just in case that the user used some unexpected input, maybe `+kubebuilder:validation:Enum` is a better choice

File: pkg/apis/pingcap/v1alpha1/types.go (hunk 9 above)
Comment (RidRisR) [reply pkg/apis/pingcap/v1alpha1/types.go:2184]: Great idea.

File: pkg/apis/pingcap/v1alpha1/backup.go [hunk 11]
```diff
@@ -316,18 +313,60 @@ func NeedNotClean(backup *Backup) bool {
...
+			subCommand = LogResumeCommand
+		} else {
+			subCommand = LogStartCommand
+		}
+	case LogStopCommand:
+		subCommand = LogStopCommand
+	case LogPauseCommand:
+		subCommand = LogPauseCommand
+	default:
+		return LogUnknownCommand
```
Comment (csuzhangxc) [inline pkg/apis/pingcap/v1alpha1/backup.go:344]: if set `LogSubcommand` as `""`, what the subcommand will be used, `log-unkown` or `log-start`.

Comment (RidRisR) [reply pkg/apis/pingcap/v1alpha1/backup.go]: Let me make it more clear:

1. 
Actually we don't need the "" as a subcommand. The only reason we keep it is for compatible with the old version interface.
The old usage is demonstrated here: https://github.com/pingcap/tidb-operator/pull/4682
TL;DR
In the old version, we use this CRD to start a new task:

```
Spec:
# empty field
```

Therefore, `LogSubcommand == ""` means `LogStartCommand`

2.
I add `LogUnknownCommand` for handling unexpected input from users, but now we use `+kubebuilder:validation:Enum` instead. This field is useless now. I keep it just for a double insurance.

3.
Now I refactored the code like

```
if backup.Spec.LogSubcommand == "" {
        if backup.Spec.LogStop || IsLogBackupAlreadyStop(backup) {
            subCommand = LogStopCommand
        } else if IsLogBackupAlreadyPaused(backup){
            subCommand = LogResumeCommand
        } else {
			subCommand = LogStartCommand
		}
    } else {
        switch backup.Spec.LogSubcommand {
        case LogStartCommand:
            if IsLogBackupAlreadyPaused(backup) {
                subCommand = LogResumeCommand
            } else {
                subCommand = LogStartCommand
            }
        case LogStopCommand:
            subCommand = LogStopCommand
        case LogPauseCommand:
            subCommand = LogPauseCommand
        default:
            return LogUnknownCommand
        }
    }
```
I suppose it could be more clear now.

**Review by WangLe1321 — APPROVED** [review]

**Review by csuzhangxc — APPROVED** [review]
